non-fungible-token/src/lib.rs

```diff
@@ -0,0 +1,222 @@
+		/// Token existence check by owner and class ID.
+		pub TokensByOwner get(fn tokens_by_owner): double_map hasher(twox_64_concat) T::AccountId, hasher(twox_64_concat) (T::ClassId, T::TokenId) => Option<()>;
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// Create NFT(non fungible token) class
+	pub fn create_class(owner: &T::AccountId, metadata: CID, data: T::ClassData) -> Result<T::ClassId, DispatchError> {
+		let class_id = NextClassId::<T>::try_mutate(|id| -> Result<T::ClassId, DispatchError> {
+			let current_id = *id;
+			*id = id.checked_add(&One::one()).ok_or(Error::<T>::NoAvailableClassId)?;
+			Ok(current_id)
+		})?;
+
+		let info = ClassInfo {
+			metadata,
+			total_issuance: Default::default(),
+			owner: owner.clone(),
+			data,
+		};
+		Classes::<T>::insert(class_id, info);
+
+		Ok(class_id)
+	}
+
+	/// Transfer NFT(non fungible token) from `from` account to `to` account
+	pub fn transfer(from: &T::AccountId, to: &T::AccountId, token: (T::ClassId, T::TokenId)) -> DispatchResult {
+		if from == to {
+			return Ok(());
+		}
+
+		TokensByOwner::<T>::try_mutate_exists(from, token, |token_by_owner| -> DispatchResult {
+			ensure!(token_by_owner.take().is_some(), Error::<T>::NoPermission);
+			TokensByOwner::<T>::insert(to, token, ());
+			Ok(())
+		})?;
+
+		Tokens::<T>::try_mutate_exists(token.0, token.1, |token_info| -> DispatchResult {
+			let mut info = token_info.as_mut().ok_or(Error::<T>::TokenNotFound)?;
+			info.owner = to.clone();
+			Ok(())
+		})
+	}
+
+	/// Mint NFT(non fungible token) to `owner`
+	pub fn mint(
+		owner: &T::AccountId,
+		class_id: T::ClassId,
+		metadata: CID,
+		data: T::TokenData,
+	) -> Result<T::TokenId, DispatchError> {
+		let token_id = NextTokenId::<T>::try_mutate(|id| -> Result<T::TokenId, DispatchError> {
+			let current_id = *id;
+			*id = id.checked_add(&One::one()).ok_or(Error::<T>::NoAvailableTokenId)?;
+			Ok(current_id)
+		})?;
+
+		Classes::<T>::try_mutate(class_id, |class_info| -> DispatchResult {
+			let info = class_info.as_mut().ok_or(Error::<T>::ClassNotFound)?;
+			info.total_issuance = info
+				.total_issuance
+				.checked_add(&One::one())
+				.ok_or(Error::<T>::NumOverflow)?;
+			Ok(())
+		})?;
+
+		let token_info = TokenInfo {
+			metadata,
+			owner: owner.clone(),
+			data,
+		};
+		Tokens::<T>::insert(class_id, token_id, token_info);
+		TokensByOwner::<T>::insert(owner, (class_id, token_id), ());
+
+		Ok(token_id)
+	}
+
+	/// Burn NFT(non fungible token) from `owner`
+	pub fn burn(owner: &T::AccountId, token: (T::ClassId, T::TokenId)) -> DispatchResult {
+		Tokens::<T>::try_mutate_exists(token.0, token.1, |token_info| -> DispatchResult {
+			ensure!(token_info.take().is_some(), Error::<T>::TokenNotFound);
+			Ok(())
+		})?;
+		TokensByOwner::<T>::try_mutate_exists(owner, token, |info| -> DispatchResult {
```

Comment:
This could fail and Tokens still get updates.

Should put rest of the method into the `Tokens:: try_mutate_exists` closure